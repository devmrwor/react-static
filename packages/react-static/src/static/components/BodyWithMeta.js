import React from 'react'
import jsesc from 'jsesc'
import { pathJoin, makePathAbsolute } from '../../utils'

const generateRouteInformation = embeddedRouteInfo => ({
  __html: `
    window.__routeInfo = JSON.parse(${jsesc(JSON.stringify(embeddedRouteInfo), {
      isScriptContext: true,
      wrap: true,
      json: true,
    })});`,
})

// Not only do we pass react-helmet attributes and the app.js here, but
// we also need to  hard code site props and route props into the page to
// prevent flashing when react mounts onto the HTML.
const makeBodyWithMeta = async state => {
  const { head, route, embeddedRouteInfo, clientScripts = [] } = state

  // This embeddedRouteInfo will be inlined into the HTML for this route.
  // It should only include the full props, not the partials.

  return ({ children, ...rest }) => (
    <body {...head.bodyProps} {...rest}>
      {children}
      {!route.redirect ? (
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={generateRouteInformation(embeddedRouteInfo)}
        />
      ) : null}
      {!route.redirect
        ? clientScripts.map(script => (
            <script
              key={script}
              defer
              type="text/javascript"
              src={makePathAbsolute(
                pathJoin(process.env.REACT_STATIC_ASSETS_PATH, script)
              )}
            />
          ))
        : null}
    </body>
  )
}

export default makeBodyWithMeta
