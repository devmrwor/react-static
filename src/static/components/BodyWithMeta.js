import React from 'react'

const REGEX_FOR_SCRIPT = /<(\/)?(script)/gi

const generateRouteInformation = embeddedRouteInfo => ({
  __html: `
    window.__routeInfo = ${JSON.stringify(embeddedRouteInfo).replace(
    REGEX_FOR_SCRIPT,
    '<"+"$1$2'
  )};`,
})

// Not only do we pass react-helmet attributes and the app.js here, but
// we also need to  hard code site props and route props into the page to
// prevent flashing when react mounts onto the HTML.
export const makeBodyWithMeta = ({
  head,
  route,
  // This embeddedRouteInfo will be inlined into the HTML for this route.
  // It should only include the full props, not the partials.
  embeddedRouteInfo,
  clientScripts = [],
  ClientCssHash,
  config,
}) => ({ children, ...rest }) => (
  <body {...head.bodyProps} {...rest}>
    {children}
    <ClientCssHash />
    {!route.redirect && (
      <script
        type="text/javascript"
        dangerouslySetInnerHTML={generateRouteInformation(embeddedRouteInfo)}
      />
    )}
    {!route.redirect &&
      clientScripts.map(script => (
        <script key={script} defer type="text/javascript" src={`${config.publicPath}${script}`} />
      ))}
  </body>
)
