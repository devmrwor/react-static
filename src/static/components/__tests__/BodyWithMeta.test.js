import React from 'react'
import { mount } from 'enzyme'
import { makeBodyWithMeta } from '../BodyWithMeta'

describe('BodyWithMeta', () => {
  test('when route is a static route', () => {
    const BodyWithMeta = makeBodyWithMeta({
      head: { bodyProps: { lang: 'en' } },
      route: { redirect: false },
      embeddedRouteInfo: {
        routeDate: 'here',
      },
      clientScripts: ['main.js', 'bootstrap.js'],
      config: {
        publicPath: 'public/path',
      },
    })

    const bodyWithMeta = mount(
      <BodyWithMeta className="body">
        <div>static page</div>
      </BodyWithMeta>
    )

    expect(bodyWithMeta).toMatchSnapshot()
  })

  test('when route is a redirect route', () => {
    const BodyWithMeta = makeBodyWithMeta({
      head: { bodyProps: { lang: 'en' } },
      route: { redirect: true },
      embeddedRouteInfo: {
        routeDate: 'here',
      },
      clientScripts: ['main.js', 'bootstrap.js'],
      config: {},
    })

    const bodyWithMeta = mount(
      <BodyWithMeta className="body">
        <div>static page</div>
      </BodyWithMeta>
    )

    expect(bodyWithMeta).toMatchSnapshot()
  })
})
