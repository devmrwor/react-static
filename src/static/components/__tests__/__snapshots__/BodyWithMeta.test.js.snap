// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyWithMeta when route is a redirect route 1`] = `
<Component
  className="body"
>
  <body
    className="body"
    lang="en"
  >
    <div>
      static page
    </div>
    <ClientCssHash>
      <style>
        body{ width:100%; height: 100%; }
      </style>
    </ClientCssHash>
  </body>
</Component>
`;

exports[`BodyWithMeta when route is a static route 1`] = `
<Component
  className="body"
>
  <body
    className="body"
    lang="en"
  >
    <div>
      static page
    </div>
    <ClientCssHash>
      <style>
        body{ width:100%; height: 100%; }
      </style>
    </ClientCssHash>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
    window.__routeInfo = {'routeDate':'here'};",
        }
      }
      type="text/javascript"
    />
    <script
      defer={true}
      key="[object Object]"
      src="undefined[object Object]"
      type="text/javascript"
    />
    <script
      defer={true}
      key="[object Object]"
      src="undefined[object Object]"
      type="text/javascript"
    />
  </body>
</Component>
`;
